compiler/src/dotty/tools/dotc/transform/PostTyper.scala

```diff
@@ -534,5 +534,25 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
         sym.addAnnotation(Annotation(defn.ExperimentalAnnot, sym.span))
         sym.companionModule.addAnnotation(Annotation(defn.ExperimentalAnnot, sym.span))
 
+    private def scala2LibPatch(tree: TypeDef)(using Context) =
+      val sym = tree.symbol
+      if ctx.settings.Yscala2Stdlib.value
+        && sym.isClass && sym.is(Module) && !sym.derivesFrom(defn.SerializableClass)
+        && sym.companionClass.derivesFrom(defn.SerializableClass)
+      then
+        // Add Serializable to companion objects of serializable classes
+        tree.rhs match
+          case impl: Template =>
+            val parents1 = impl.parents :+ TypeTree(defn.SerializableType)
+            val impl1 = cpy.Template(impl)(parents = parents1)
+            cpy.TypeDef(tree)(rhs = impl1)
+      else tree
   }
+
+  protected override def infoMayChange(sym: Symbol)(using Context): Boolean = ctx.settings.Yscala2Stdlib.value
+
+  def transformInfo(tp: Type, sym: Symbol)(using Context): Type = tp match
+    case info: ClassInfo if sym.is(Module, butNot = Package) && !sym.derivesFrom(defn.SerializableClass) =>
+      info.derivedClassInfo(declaredParents = info.parents :+ defn.SerializableType)
```

Comment:
This is missing the `sym.companionClass.derivesFrom(defn.SerializableClass)` test.